### ISSUE TITLE
Authentication user provider [adldap] is not defined

### ISSUE THREAD

**Issue by faaazma** [Adldap2/Adldap2-Laravel#44]
I already change provider  for web and add the provider in config\auth.php as instructed but getting the error.
```php
ErrorException in CreatesUserProviders.php line 40:
Authentication user provider [adldap] is not defined. (View: C:\laravel5\blog\resources\views\layouts\app.blade.php) (View: C:\laravel5\blog\resources\views\layouts\app.blade.php)
```

Please advise.
i already add Adldap\Laravel\AdldapServiceProvider::class and the facade to the app.php

**Comment by stevebauman** [comment]
Did you add the `Adldap\Laravel\AdldapAuthServiceProvider::class` to your `config/app.php` file?

**Comment by faaazma** [comment]
Yes I did as below.
![config_app_php](https://cloud.githubusercontent.com/assets/12085378/12413390/765c110c-bec9-11e5-9bc3-dd633b0411ea.jpg)

**Comment by faaazma** [comment]
However i put "adldap2/adldap2-laravel": "1.4.*" under "require-dev". should i put it under "require" in composer.json?

**Comment by stevebauman** [comment]
You haven't inserted the `AdldapAuthServiceProvider` in your `config/app.php` file according to your attached image.

You need both providers:

```php
Adldap\Laravel\AdldapServiceProvider::class,
Adldap\Laravel\AdldapAuthServiceProvider::class
```

Also for your other question, if your project requires Adldap in production, you will need it in the `require` section of your `composer.json` file.

**Comment by faaazma** [comment]
Thanks Steve. It is working after i add the Adldap\Laravel\AdldapAuthServiceProvider::class.
I guess I missed it since it is not in your intro writeup.

**Comment by stevebauman** [comment]
No problem. It's actually located in the [Auth driver setup](https://github.com/Adldap2/Adldap2-Laravel#auth-driver).

**Comment by iraklisg** [comment]
@stevebauman @faaazma I faced the same problem running `L5.5` with `adldap 2.1.3` via **docker**.
For some peculiar reason, adding the service providers in `app.php` didn't work out (although it works in other laravel application that I am running (but not via docker)) 

In addition, running `php artisan package:discover` returned all the packages that I have added in `app.php` _except_ the adldap2 packages.

What I did to overcome the problem was to register the adlap providers in `AppServiceProvider.php`

```
public function register()
{
    $this->app->register('Adldap\Laravel\AdldapServiceProvider');
    $this->app->register('Adldap\Laravel\AdldapAuthServiceProvider');
}
```
When I did that the problem gone :confused:

**Comment by stevebauman** [comment]
Hi @iraklisg,

> In addition, running php artisan package:discover returned all the packages that I have added in app.php except the adldap2 packages.

This is because `v3.0` doesn't include the package discovery feature that was released with Laravel 5.5, which auto-registers the packages service providers for you. So you will have to register them manually for now.

`v4.0` will have this feature included, but I will also update the `v3.0` branch to include this as well.

**Comment by stevebauman** [comment]
Added: https://github.com/Adldap2/Adldap2-Laravel/releases/tag/v3.0.7

**Comment by oksim14** [comment]
Hi, same issue.

adldap2/adldap2                       v8.1.3
adldap2/adldap2-laravel               v4.0.6

Config/auth.php
```     
    'providers' => [
        'users' => [
            'driver' => 'ldap',
            'model' => App\Models\User::class,
        ],...
```

Config/adldap_auth.php published

And still this error:

```
InvalidArgumentException
Authentication user provider [ldap] is not defined.
```

```
/vendor/laravel/framework/src/Illuminate/Auth/CreatesUserProviders.php
     */
    public function createUserProvider($provider = null)
    {
        if (is_null($config = $this->getProviderConfiguration($provider))) {
            return;
        }
 
        if (isset($this->customProviderCreators[$driver = ($config['driver'] ?? null)])) {
            return call_user_func(
                $this->customProviderCreators[$driver], $this->app, $config
            );
        }
 
        switch ($driver) {
            case 'database':
                return $this->createDatabaseProvider($config);
            case 'eloquent':
                return $this->createEloquentProvider($config);
            default:
                throw new InvalidArgumentException(
                    "Authentication user provider [{$driver}] is not defined."
                );
        }
    }
```

`"Authentication user provider [ldap] is not defined."`

**Comment by stevebauman** [comment]
Hi @oksim14, your configuration is incorrect.

The `driver` name should be `adldap`, not `ldap`.

**Comment by oksim14** [comment]
🥇 Thanks a lot... Love these 'bugs' :)

**Comment by michaelcmundo** [comment]
Hi I'm using laravel 5.7 and I got Authentication user provider [adldap] is not defined.

I added the following to AppServiceProvider.php inside register()
        $this->app->register('Adldap\Laravel\AdldapServiceProvider');
        $this->app->register('Adldap\Laravel\AdldapAuthServiceProvider');
And also tried in app.php with the following in provider array:
        Adldap\Laravel\AdldapServiceProvider::class,
        Adldap\Laravel\AdldapAuthServiceProvider::class,
And 'Adldap' => Adldap\Laravel\Facades\Adldap::class, in aliases array 

I have 'provider' => Adldap\Laravel\Auth\NoDatabaseUserProvider::class, in ldap_auth.php

but still not work. Any idea what's the issue?


![laravel-err](https://user-images.githubusercontent.com/7003801/48445786-77a7d100-e765-11e8-8606-ee91179e13fb.PNG)

**Comment by liqSTAR** [comment]
Same problem. Installed it today and got the same error and i dont find the right the error.

**Comment by stevebauman** [comment]
Hi @michaelachrisco, the driver name is 'ldap', not 'adldap'.

Open your `config/auth.php` file and change your `driver` option to `ldap`.

This is described in the upgrade guide:

https://adldap2.github.io/Adldap2-Laravel/#/upgrading?id=authentication-driver

**Comment by michaelachrisco** [comment]
@stevebauman I believe you meant @michaelcmundo not me. But its all good 👍

**Comment by stevebauman** [comment]
Shoot sorry about that @michaelachrisco! Should have checked the auto-correct 😅

**Comment by michaelcmundo** [comment]
Hi I changed to ldap and now I got this error instead. Any idea?

![ldap-err](https://user-images.githubusercontent.com/7003801/48864483-aee83480-ed9a-11e8-880f-488c8b963d48.PNG)

**Comment by stevebauman** [comment]
@michaelcmundo Is the input name of the username field in your `login.blade.php` file `userprincipalname`? Or is it something else?

**Comment by michaelcmundo** [comment]
In my login.blade.php, the name is username. I just noticed in config/ldap_auth.php that the usernames array has these:
        'ldap' => [
            'discover' => 'userprincipalname',
            'authenticate' => 'distinguishedname',
        ]

Maybe my issue is because of these? What are those?

**Comment by michaelcmundo** [comment]
I just changed those 2 to have values 'username' as I don't really know what it is. Before I use laravel I just use php ldap_connect() and ldap_bind(). I don't need to specify the name of the username field, etc.

**Comment by michaelcmundo** [comment]
Yea I read it after but I don't know as previously I just use ldap_connect() and ldap_bind() which doesn't need to set those. So I tried change the config to have username instead of userprincipalname and when I tried log in I don't get the error but I always getting "These credentials do not match our records". The credentials I put in are correct but just not sure the exact name.

Is there a way I can turn on debug or logging so I can see exact name/values it's passing to the ldap?
